File: lingxia-webview/android/lingxia/src/main/java/com/lingxia/lxapp/NavigationBar.kt
package com.lingxia.miniapp

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.util.Log
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.annotation.AttrRes
import androidx.annotation.StyleRes
import org.json.JSONObject
import android.view.animation.AccelerateDecelerateInterpolator
import kotlin.math.max
import kotlin.math.min

/**
 * Configuration data class for the NavigationBar
 */
data class NavigationBarConfig(
    val hidden: Boolean = false,                   // Whether the navigation bar is hidden
    val navigationBarBackgroundColor: Int? = null, // Background color (e.g., #FFFFFF)
    val navigationBarTextStyle: String? = null,    // Text style ("black" or "white")
    val navigationBarTitleText: String? = null,    // Navigation bar title text
    val navigationStyle: String? = null            // "default" or "custom"
) {
    companion object {
        // Default values
        val DEFAULT_BACKGROUND_COLOR = Color.WHITE
        val DEFAULT_TEXT_COLOR = Color.BLACK
        const val DEFAULT_HEIGHT_DP = LxAppActivity.DEFAULT_NAV_BAR_HEIGHT_DP

        fun fromJson(json: String?): NavigationBarConfig? {
            if (json.isNullOrEmpty()) {
                // Return default config if JSON is missing/empty, assuming NavBar should exist but be hidden
                return NavigationBarConfig(hidden = true)
            }

            return try {
                val jsonObject = JSONObject(json)

                // Handle navigation style - if "custom", we should hide the standard nav bar
                val navStyle = jsonObject.optString("navigationStyle", "default")
                val isHidden = jsonObject.optBoolean("hidden", false) || navStyle == "custom"

                // Parse text style (black or white)
                val textStyle = jsonObject.optString("navigationBarTextStyle", "black")

                NavigationBarConfig(
                    hidden = isHidden,
                    navigationBarBackgroundColor = parseColor(jsonObject.optString("navigationBarBackgroundColor"), DEFAULT_BACKGROUND_COLOR),
                    navigationBarTextStyle = textStyle,
                    navigationBarTitleText = jsonObject.optString("navigationBarTitleText", ""),
                    navigationStyle = navStyle
                )
            } catch (e: Exception) {
                Log.e("NavBarConfig", "Error parsing NavigationBar config: ${e.message}")
                NavigationBarConfig(hidden = true) // Fallback to hidden on parse error
            }
        }

        private fun parseColor(colorString: String?, defaultColor: Int): Int {
            if (colorString.isNullOrEmpty()) return defaultColor
            return try {
                Color.parseColor(colorString)
            } catch (e: Exception) {
                Log.w("NavBarConfig", "Invalid color string '$colorString'. Using default.")
                defaultColor
            }
        }
    }
}

/**
 * Custom Navigation Bar view mimicking WeChat Mini Program behavior.
 */
class NavigationBar @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    @AttrRes defStyleAttr: Int = 0,
    @StyleRes defStyleRes: Int = 0
) : FrameLayout(context, attrs, defStyleAttr, defStyleRes) {

    companion object {
        private const val TAG = "LingXia.NavigationBar"
        // Default colors - internal access
        internal val DEFAULT_BACKGROUND_COLOR = Color.WHITE
        internal val DEFAULT_FRONT_COLOR = Color.BLACK // Default text/icon color

        // Define a specific height for tablets
        private const val DEFAULT_TABLET_HEIGHT_DP = 12
    }

    private val titleTextView: TextView
    private val loadingIndicator: ProgressBar
    private val backButton: ImageView
    private val homeButton: ImageView? = null
    private var currentConfig: NavigationBarConfig = NavigationBarConfig()
    private var knownStatusBarHeight: Int = 0

    // Store current colors
    private var currentBackgroundColor = DEFAULT_BACKGROUND_COLOR
    private var currentFrontColor = DEFAULT_FRONT_COLOR

    // Callbacks
    private var onBackClickListener: (() -> Unit)? = null

    /**
     * Custom back button drawable that draws a chevron "<" shape
     */
    private inner class BackButtonDrawable : Drawable() {
        private val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            color = currentFrontColor
            style = Paint.Style.STROKE
            strokeWidth = 1.8f * resources.displayMetrics.density
            strokeCap = Paint.Cap.ROUND
            strokeJoin = Paint.Join.ROUND
        }

        override fun draw(canvas: Canvas) {
            val width = bounds.width()
            val height = bounds.height()
            val centerY = height / 2f
            val startX = width * 0.55f
            val endX = width * 0.35f

            // Draw the chevron lines
            canvas.drawLine(startX, centerY - height * 0.15f, endX, centerY, paint)
            canvas.drawLine(endX, centerY, startX, centerY + height * 0.15f, paint)
        }

        override fun setAlpha(alpha: Int) {
            paint.alpha = alpha
        }

        override fun setColorFilter(colorFilter: android.graphics.ColorFilter?) {
            paint.colorFilter = colorFilter
        }

        @Deprecated("Deprecated in Java")
        override fun getOpacity(): Int = android.graphics.PixelFormat.TRANSLUCENT

        fun updateColor(color: Int) {
            paint.color = color
            invalidateSelf()
        }
    }

    init {
        val density = resources.displayMetrics.density

        // Determine if it's a tablet (smallest width >= 600dp)
        val smallestScreenWidthDp = context.resources.configuration.smallestScreenWidthDp
        val isTablet = smallestScreenWidthDp >= 600

        val navBarHeightDp = if (isTablet) DEFAULT_TABLET_HEIGHT_DP else LxAppActivity.DEFAULT_NAV_BAR_HEIGHT_DP

        Log.d(TAG, "smallestScreenWidthDp: $smallestScreenWidthDp, isTablet: $isTablet, navBarHeightDp: $navBarHeightDp")
        val heightPx = (navBarHeightDp * density).toInt()

        setBackgroundColor(currentBackgroundColor)

        // Back Button setup
        backButton = ImageView(context).apply {
            layoutParams = LayoutParams(heightPx, heightPx).apply {
                gravity = Gravity.START or Gravity.TOP
                marginStart = (4 * density).toInt()
            }
            setImageDrawable(BackButtonDrawable())
            contentDescription = "Back"
            visibility = View.GONE
        }
        addView(backButton)

        // Calculate dynamic font size for title
        val targetTitleSp = if (isTablet) 12f else 17f

        Log.d(TAG, "Device isTablet: $isTablet, navBarHeightDp: $navBarHeightDp, Setting title font size to: $targetTitleSp sp")

        // Title TextView setup
        titleTextView = TextView(context).apply {
            layoutParams = LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT).apply {
                gravity = Gravity.CENTER_HORIZONTAL or Gravity.TOP
                // Align title with capsule button center: 24dp (status bar) + 18dp (adjustment)
                topMargin = (42 * density).toInt()
            }
            gravity = Gravity.CENTER
            textAlignment = View.TEXT_ALIGNMENT_CENTER
            setTextColor(currentFrontColor)
            includeFontPadding = false
            setTextSize(TypedValue.COMPLEX_UNIT_SP, targetTitleSp)
            typeface = android.graphics.Typeface.create("sans-serif-medium", android.graphics.Typeface.NORMAL)
            visibility = View.VISIBLE
        }
        addView(titleTextView)

        // Loading Indicator setup
        val progressBarSize = (24 * density).toInt()
        loadingIndicator = ProgressBar(context, null, android.R.attr.progressBarStyleSmall).apply {
            layoutParams = LayoutParams(progressBarSize, progressBarSize).apply {
                gravity = Gravity.CENTER_VERTICAL or Gravity.START
                marginStart = (16 * density).toInt()
            }
            updateProgressColor(currentFrontColor)
            visibility = View.GONE
        }
        addView(loadingIndicator)
    }

    // Helper method to update progress indicator color
    private fun ProgressBar.updateProgressColor(color: Int) {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.Q) {
            indeterminateDrawable?.colorFilter = android.graphics.BlendModeColorFilter(color, android.graphics.BlendMode.SRC_IN)
        } else {
            @Suppress("DEPRECATION")
            indeterminateDrawable?.setColorFilter(color, android.graphics.PorterDuff.Mode.SRC_IN)
        }
    }

    // Keep these layout debugging overrides only in debug builds
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        Log.d(TAG, "onMeasure: widthSpec=${MeasureSpec.toString(widthMeasureSpec)}, heightSpec=${MeasureSpec.toString(heightMeasureSpec)}")
        Log.d(TAG, "onMeasure: measuredWidth=$measuredWidth, measuredHeight=$measuredHeight")
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        Log.d(TAG, "onLayout: changed=$changed, left=$left, top=$top, right=$right, bottom=$bottom")
        Log.d(TAG, "onLayout: width=$width, height=$height, measuredWidth=$measuredWidth, measuredHeight=$measuredHeight")
        if (titleTextView.visibility == View.VISIBLE) {
            Log.d(TAG, "onLayout: titleTextView.top=${titleTextView.top}, titleTextView.bottom=${titleTextView.bottom}, titleTextView.height=${titleTextView.height}, titleTextView.measuredHeight=${titleTextView.measuredHeight}")
        }
    }

    /**
     * Returns the calculated content height in pixels based on device type (phone/tablet).
     */
    fun getCalculatedContentHeightPx(): Int {
        val density = resources.displayMetrics.density
        val isTablet = context.resources.configuration.smallestScreenWidthDp >= 600
        val navBarHeightDp = if (isTablet) DEFAULT_TABLET_HEIGHT_DP else LxAppActivity.DEFAULT_NAV_BAR_HEIGHT_DP
        return (navBarHeightDp * density).toInt()
    }

    /**
     * Shows the loading indicator in the navigation bar.
     */
    fun showLoading() {
        loadingIndicator.visibility = View.VISIBLE
    }

    /**
     * Hides the loading indicator in the navigation bar.
     */
    fun hideLoading() {
        loadingIndicator.visibility = View.GONE
    }

    /**
     * Sets the title text displayed in the navigation bar.
     *
     * @param title The text to display.
     */
    fun setTitle(title: String?) {
        titleTextView.text = title ?: ""
    }

    /**
     * Sets the background color and front color (for title and loading indicator) of the navigation bar.
     *
     * @param backgroundColor The background color (e.g., Color.parseColor("#ffffff")).
     * @param frontColor The color for text and icons (e.g., Color.parseColor("#000000")).
     */
    fun setColor(backgroundColor: Int, frontColor: Int) {
        currentBackgroundColor = backgroundColor
        currentFrontColor = frontColor

        setBackgroundColor(currentBackgroundColor)
        titleTextView.setTextColor(currentFrontColor)

        // Update loading indicator color
        loadingIndicator.updateProgressColor(currentFrontColor)

        // Update back button color
        (backButton.drawable as? BackButtonDrawable)?.updateColor(currentFrontColor)
    }

    /**
     * Sets the visibility of the back button.
     *
     * @param visible Whether the back button should be visible.
     */
    fun setBackButtonVisible(visible: Boolean) {
        backButton.visibility = if (visible) View.VISIBLE else View.GONE
    }

    /**
     * Sets a listener for back button clicks
     *
     * @param listener The callback to invoke when the back button is clicked
     */
    fun setOnBackButtonClickListener(listener: OnClickListener) {
        backButton.setOnClickListener(listener)
    }

    /**
     * Hides the navigation bar.
     */
    fun hide() {
        visibility = View.GONE
    }

    /**
     * Updates the state of the NavigationBar and optionally animates the transition.
     *
     * @param title The title text to display.
     * @param bgColor The background color.
     * @param textColor The text color (for title and button icons).
     * @param showBackButton Whether the back button should be initially visible.
     * @param isBackNavigation Direction hint for animation.
     * @param disableAnimation If true, update instantly; otherwise, animate.
     * @param onBackClickListener Listener for the back button.
     * @param onAnimationEnd Optional Runnable to execute after animation finishes.
     */
    fun updateStateAndAnimate(
        title: String,
        bgColor: Int,
        textColor: Int,
        showBackButton: Boolean,
        isBackNavigation: Boolean,
        disableAnimation: Boolean,
        onBackClickListener: OnClickListener,
        onAnimationEnd: Runnable? = null
    ) {
        visibility = View.VISIBLE

        // Set state
        setTitle(title)
        setColor(bgColor, textColor)
        setBackButtonVisible(showBackButton)
        setOnBackButtonClickListener(onBackClickListener)

        // Handle animation
        if (!disableAnimation) {
            val animStartX = if (isBackNavigation) -width.toFloat() else width.toFloat()
            val duration = 250L

            translationX = animStartX

            animate()
                .translationX(0f)
                .setDuration(duration)
                .setInterpolator(AccelerateDecelerateInterpolator())
                .withEndAction { // Use the provided end action
                    translationX = 0f // Ensure final position
                    onAnimationEnd?.run() // Execute the callback
                }
                .start()
        } else {
            translationX = 0f
            // If no animation, run the end action immediately if it exists,
            // as it might contain layout updates needed right away.
            onAnimationEnd?.run()
        }
    }

    // Method to receive status bar height
    fun setExternalStatusBarHeight(sbh: Int) {
        if (knownStatusBarHeight != sbh) {
            Log.d(TAG, "ExternalStatusBarHeight set to: $sbh")
            knownStatusBarHeight = sbh

            // Update layout params of children that depend on status bar height
            listOf(backButton, loadingIndicator).forEach { view ->
                (view.layoutParams as? FrameLayout.LayoutParams)?.let {
                    it.topMargin = knownStatusBarHeight
                    view.layoutParams = it
                }
            }

            requestLayout() // Request re-layout
        }
    }
}

File: lingxia-webview/ios/lingxia/Sources/NavigationBar.swift
import UIKit
import Foundation
import os.log

/// Configuration data class for the NavigationBar
public struct NavigationBarConfig {
    let hidden: Bool
    let navigationBarBackgroundColor: UIColor?
    let navigationBarTextStyle: String?
    let navigationBarTitleText: String?
    let navigationStyle: String?

    static let DEFAULT_BACKGROUND_COLOR = UIColor.white
    static let DEFAULT_TEXT_COLOR = UIColor.black
    static let DEFAULT_HEIGHT: CGFloat = 44

    public init(
        hidden: Bool = false,
        navigationBarBackgroundColor: UIColor? = nil,
        navigationBarTextStyle: String? = nil,
        navigationBarTitleText: String? = nil,
        navigationStyle: String? = nil
    ) {
        self.hidden = hidden
        self.navigationBarBackgroundColor = navigationBarBackgroundColor
        self.navigationBarTextStyle = navigationBarTextStyle
        self.navigationBarTitleText = navigationBarTitleText
        self.navigationStyle = navigationStyle
    }

    public static func fromJson(_ json: String?) -> NavigationBarConfig? {
        guard let json = json, !json.isEmpty else {
            return NavigationBarConfig(hidden: true)
        }

        do {
            guard let data = json.data(using: .utf8),
                  let jsonObject = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                return NavigationBarConfig(hidden: true)
            }

            let navStyle = jsonObject["navigationStyle"] as? String ?? "default"
            let isHidden = (jsonObject["hidden"] as? Bool ?? false) || navStyle == "custom"
            let textStyle = jsonObject["navigationBarTextStyle"] as? String ?? "black"

            return NavigationBarConfig(
                hidden: isHidden,
                navigationBarBackgroundColor: parseColor(jsonObject["navigationBarBackgroundColor"] as? String, defaultColor: DEFAULT_BACKGROUND_COLOR),
                navigationBarTextStyle: textStyle,
                navigationBarTitleText: jsonObject["navigationBarTitleText"] as? String ?? "",
                navigationStyle: navStyle
            )
        } catch {
            return NavigationBarConfig(hidden: true)
        }
    }

    private static func parseColor(_ colorString: String?, defaultColor: UIColor) -> UIColor {
        guard let colorString = colorString, !colorString.isEmpty else { return defaultColor }

        if colorString.hasPrefix("#") {
            return UIColor(hexString: colorString) ?? defaultColor
        }
        return defaultColor
    }
}

/// Custom Navigation Bar view for LxApp system
public class LingXiaNavigationBar: UIView {
    private static let log = OSLog(subsystem: "LingXia", category: "NavigationBar")

    internal static let DEFAULT_BACKGROUND_COLOR = UIColor.white
    internal static let DEFAULT_FRONT_COLOR = UIColor.black
    private static let DEFAULT_TABLET_HEIGHT: CGFloat = 44

    private let titleLabel: UILabel
    private let loadingIndicator: UIActivityIndicatorView
    private let backButton: UIButton
    private var currentConfig: NavigationBarConfig = NavigationBarConfig()
    private var knownStatusBarHeight: CGFloat = 0

    private var currentBackgroundColor = DEFAULT_BACKGROUND_COLOR
    private var currentFrontColor = DEFAULT_FRONT_COLOR

    private var onBackClickListener: (() -> Void)?

    public override init(frame: CGRect) {
        titleLabel = UILabel()
        loadingIndicator = UIActivityIndicatorView(style: .medium)
        backButton = UIButton(type: .custom)

        super.init(frame: frame)
        setupUI()
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    private func setupUI() {
        backgroundColor = currentBackgroundColor

        let isTablet = UIDevice.current.userInterfaceIdiom == .pad
        let _ = isTablet ? LingXiaNavigationBar.DEFAULT_TABLET_HEIGHT : LxAppViewController.DEFAULT_NAV_BAR_HEIGHT

        backButton.setTitle("‹", for: .normal)
        backButton.setTitleColor(currentFrontColor, for: .normal)
        backButton.titleLabel?.font = UIFont.systemFont(ofSize: 24, weight: .medium)
        backButton.contentHorizontalAlignment = .center
        backButton.isHidden = true
        backButton.addTarget(self, action: #selector(backButtonTapped), for: .touchUpInside)

        let targetTitleSize: CGFloat = isTablet ? 12 : 17

        titleLabel.textAlignment = .center
        titleLabel.textColor = currentFrontColor
        titleLabel.font = UIFont.systemFont(ofSize: targetTitleSize, weight: .medium)
        titleLabel.numberOfLines = 1

        loadingIndicator.color = currentFrontColor
        loadingIndicator.hidesWhenStopped = true

        addSubview(backButton)
        addSubview(titleLabel)
        addSubview(loadingIndicator)

        setupConstraints()
    }

    private func setupConstraints() {
        backButton.translatesAutoresizingMaskIntoConstraints = false
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        loadingIndicator.translatesAutoresizingMaskIntoConstraints = false

        // Use safe area guide for better layout on different devices
        let _ = safeAreaLayoutGuide

        NSLayoutConstraint.activate([
            // Back button constraints
            backButton.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 4),
            backButton.topAnchor.constraint(equalTo: topAnchor, constant: LxAppViewController.NAV_TITLE_VERTICAL_POSITION - 2),
            backButton.widthAnchor.constraint(equalToConstant: 44),
            backButton.heightAnchor.constraint(equalToConstant: 32),

            // Title label constraints - ensure it doesn't overlap with back button
            titleLabel.centerXAnchor.constraint(equalTo: centerXAnchor),
            titleLabel.topAnchor.constraint(equalTo: topAnchor, constant: LxAppViewController.NAV_TITLE_VERTICAL_POSITION),
            titleLabel.leadingAnchor.constraint(greaterThanOrEqualTo: backButton.trailingAnchor, constant: 8),
            titleLabel.trailingAnchor.constraint(lessThanOrEqualTo: trailingAnchor, constant: -52), // Leave space for potential right button

            // Loading indicator constraints
            loadingIndicator.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 16),
            loadingIndicator.topAnchor.constraint(equalTo: topAnchor, constant: LxAppViewController.NAV_TITLE_VERTICAL_POSITION),
            loadingIndicator.widthAnchor.constraint(equalToConstant: 24),
            loadingIndicator.heightAnchor.constraint(equalToConstant: 24)
        ])
    }

    /// Returns the calculated content height based on device type
    public func getCalculatedContentHeight() -> CGFloat {
        let isTablet = UIDevice.current.userInterfaceIdiom == .pad
        return isTablet ? LingXiaNavigationBar.DEFAULT_TABLET_HEIGHT : LxAppViewController.DEFAULT_NAV_BAR_HEIGHT
    }

    /// Shows the loading indicator
    public func showLoading() {
        loadingIndicator.startAnimating()
    }

    /// Hides the loading indicator
    public func hideLoading() {
        loadingIndicator.stopAnimating()
    }

    /// Sets the title text
    public func setTitle(_ title: String?) {
        titleLabel.text = title ?? ""
    }

    /// Sets the background and front colors
    public func setColor(backgroundColor: UIColor, frontColor: UIColor) {
        currentBackgroundColor = backgroundColor
        currentFrontColor = frontColor

        self.backgroundColor = currentBackgroundColor
        titleLabel.textColor = currentFrontColor
        loadingIndicator.color = currentFrontColor
        backButton.setTitleColor(currentFrontColor, for: .normal)
    }

    /// Sets the visibility of the back button
    public func setBackButtonVisible(_ visible: Bool) {
        backButton.isHidden = !visible
    }

    /// Sets a listener for back button clicks
    public func setOnBackButtonClickListener(_ listener: @escaping () -> Void) {
        onBackClickListener = listener
    }

    @objc private func backButtonTapped() {
        onBackClickListener?()
    }

    /// Hides the navigation bar
    public func hide() {
        isHidden = true
    }

    /// Updates the NavigationBar with provided configuration
    /// Returns true if NavigationBar should be shown, false if it should be hidden
    public func updateWithConfig(
        pageConfig: NavigationBarConfig?,
        isBackNavigation: Bool,
        disableAnimation: Bool,
        onBackClickListener: @escaping () -> Void,
        onAnimationEnd: (() -> Void)? = nil
    ) -> Bool {
        // Check if NavigationBar should be hidden
        let shouldHide = pageConfig?.hidden ?? false
        if shouldHide {
            hide()
            return false
        }

        // Extract configuration values with defaults
        let titleText = pageConfig?.navigationBarTitleText ?? ""
        let backgroundColor = pageConfig?.navigationBarBackgroundColor ?? NavigationBarConfig.DEFAULT_BACKGROUND_COLOR
        let textStyle = pageConfig?.navigationBarTextStyle ?? "black"
        let textColor = textStyle == "white" ? UIColor.white : UIColor.black
        let showBackButton = !disableAnimation

        // Update state with provided configuration
        updateStateAndAnimate(
            title: titleText,
            bgColor: backgroundColor,
            textColor: textColor,
            showBackButton: showBackButton,
            isBackNavigation: isBackNavigation,
            disableAnimation: disableAnimation,
            onBackClickListener: onBackClickListener,
            onAnimationEnd: onAnimationEnd
        )

        return true
    }

    /// Updates the state of the NavigationBar and optionally animates the transition
    public func updateStateAndAnimate(
        title: String,
        bgColor: UIColor,
        textColor: UIColor,
        showBackButton: Bool,
        isBackNavigation: Bool,
        disableAnimation: Bool,
        onBackClickListener: @escaping () -> Void,
        onAnimationEnd: (() -> Void)? = nil
    ) {
        isHidden = false

        setTitle(title)
        setColor(backgroundColor: bgColor, frontColor: textColor)
        setBackButtonVisible(showBackButton)
        setOnBackButtonClickListener(onBackClickListener)

        if !disableAnimation {
            let animStartX: CGFloat = isBackNavigation ? -frame.width : frame.width
            let duration: TimeInterval = 0.25

            transform = CGAffineTransform(translationX: animStartX, y: 0)

            UIView.animate(withDuration: duration, animations: {
                self.transform = .identity
            }) { _ in
                self.transform = .identity
                onAnimationEnd?()
            }
        } else {
            transform = .identity
            onAnimationEnd?()
        }
    }

    /// Updates status bar height for layout calculations
    public func setExternalStatusBarHeight(_ sbh: CGFloat) {
        if knownStatusBarHeight != sbh {
            knownStatusBarHeight = sbh
            setNeedsUpdateConstraints()
        }
    }
}

// UIColor hex extension is defined in UIColor+Hex.swift
